solidity/random-beacon/contracts/RandomBeacon.sol

```diff
@@ -716,55 +717,85 @@ contract RandomBeacon is Ownable {
     }
 
     /// @notice Creates a new relay entry.
+    /// @dev This function can be called only before the soft timeout. This is a
+    ///      happy scenario without slashing.
+    /// @param entry Group BLS signature over the previous entry.
+    function submitRelayEntry(bytes calldata entry) external {
+        require(relay.isRequestInProgress(), "No relay request in progress");
+        require(
+            !relay.hasRequestSoftTimedOut(),
+            "Relay submission passed a soft timeout"
+        );
+
+        Groups.Group storage group = groups.getGroup(
+            relay.currentRequestGroupID
+        );
+
+        relay.submitEntry(entry, group.groupPubKey);
+
+        // If DKG is awaiting a seed, that means the we should start the actual
+        // group creation process.
+        if (dkg.currentState() == DKG.State.AWAITING_SEED) {
+            dkg.start(uint256(keccak256(entry)));
+        }
+
+        callback.executeCallback(uint256(keccak256(entry)), callbackGasLimit);
+    }
+
+    /// @notice Creates a new relay entry.
+    /// @dev It should be called only after a soft timeout. Group members will be
+    ///      slashed. Function reverts if a hard timeout is hit.
     /// @param entry Group BLS signature over the previous entry.
     /// @param groupMembers Group member ids that participated in dkg (excluding IA/DQ).
     function submitRelayEntry(
         bytes calldata entry,
         uint32[] calldata groupMembers
     ) external {
-        uint256 currentRequestId = relay.currentRequestID;
+        require(relay.isRequestInProgress(), "No relay request in progress");
+        require(
+            relay.hasRequestSoftTimedOut(),
+            "Relay did not pass a soft timeout"
+        );
+        require(!relay.hasRequestHardTimedOut(), "Relay request timed out");
```

Comment:
The logic inside of `hasRequestHardTimedOut()` was implemented in `hasRequestTimedOut()`. `hasRequestTimedOut()` was used in a couple of places: `cleanupOnEntryTimeout()`, `retryOnEntryTimeout()` & `submitEntry()`. The only difference between these two functions is that in `hasRequestTimedOut()` we had `isRequestInProgress()`. `hasRequestHardTimedOut()` decouples `isRequestInProgress()` and moves it upstream to `cleanupOnEntryTimeout()`, `retryOnEntryTimeout()` & `submitEntry()`.

We have two timeouts: soft and hard. Since after the split we need to check the soft timeout, I decided to rename `hasRequestTimedOut()` -> `hasRequestHardTimedOut()` cause this is essentially what we're checking there.

Reply:
Not sure if I'm fully following you Piotr. Extracting `isRequestInProgress()` from `hasRequestTimedOut()` and moving `isRequestInProgress()` upstream to these `cleanupOnEntryTimeout(), retryOnEntryTimeout() & submitEntry()` functions doesn't seem to change any "business" logic. The intention is to refactor a bit without changing the core flows. If I'm not missing something, then whatever concerns we have if we allow submitting the entry after the hard timeout, they also present in the `main` branch as well.